### DISCUSSION TITLE
左右およびトラックボールの有無の自動判定する改造

### DISCUSSION THREAD

**Discussion by koron** [Yowkees/keyball#16] [Ideas]
QMKには[分割キーボードの左右を自動判定させる機能](https://beta.docs.qmk.fm/using-qmk/hardware-features/feature_split_keyboard#setting-handedness)があります。
現在の Keyball はUSBをどちらにつないだかで左右の判定をしていますが
適切なジャンパーを付ける改造をすれば
QMKのこの機能を使うことで左右を正しく自動判定できます(そういうファームを書けます)。

(ファームで特定のジャンパがあるかどうかをみて、左右を判断するということです)

またトラックボールについても同様で、
ファームから見てトラックボールがあるかどうか知る手段があれば
USBをどちらにつないでもトラックボールが機能する余地が生まれ
使い勝手が向上すると考えられます。
またダブルトラックボール構想の実現も視野に入ります。
こちらはSPIで光学センサーの有無を調べられればわかりそうですが、
別解として左右判定と同様にジャンパー改造も選択肢になりえます。

手元の機体は適当に改造してしまえば何とでもなるのですが
その場合でもファームウェアはいじるので
せっかくですから設計者の Yowkees さんの考えと擦り合わせられないかな
と考えています。

なにかアイデアとか意向とかありますでしょうか?

**Comment by Yowkees** [comment]
ご提案ありがとうございます！
いやー知らないことばかりで恐縮です。次回のロットよりジャンパーを追加するようにします。
実は左手基板には表裏を判別するプルダウンのジャンパーを設けておいたのですが、ダブルトラックボール仕様では無意味ですもんね。
少しだけ左右判別の説明を読んでみました。
現状右手基板には空きピンはありませんが（すみません。まだ回路図をアップしていませんでした。公開するようにします）、マトリクスピンでの読み取りというやつでできるんでしょうかね。もっと熟読する必要がありそうです。お時間ください。
それか、こことここの間にジャンパーと指示いただければそのように修正します。ソフト製作上有利な方法教えていただければと思います。
有益な情報と、多大なるご協力、誠にありがとうございます。

**Reply by koron** [reply]
了解です。左右判定はマトリックスでもできます。
空いてるマトリクスに1つダイオード入れてあげて、
あとはマトリクスマスクを設定してあげれば完了です。

マトリックスマスクは config.h で `#define MATRIX_MASKED` して
以下のような感じで `matrix_mask` を定義してあげればOKです。
LSBが第1カラムなので左右が逆転しているのですが
キーの場所は `1` にしてジャンパー(≒ダイオードだけ)のところは `0` にします。
こうするとオンになっていてもキーが押されたとは判定されないので
もろもろの内部状態の更新が避けられます。

```
matrix_row_t matrix_mask[MATRIX_ROWS] = {
    0b111111111111111,
    0b101111111111111,
    0b100111111111111,
    0b110111111111101,
    0b000110001000110,
};
```

(ちなみに上のマトリックスは自作のHHKBレイアウトのやつ)

**Comment by koron** [comment]
左右自動判定は、左側用とするボードの `SW` をショートすることで行うことに
https://github.com/Yowkees/keyball/discussions/24 で決まりました。
